src/hotspot/share/runtime/handshake.cpp

```diff
@@ -465,60 +443,78 @@ bool HandshakeState::possibly_can_process_handshake() {
   }
 }
 
-bool HandshakeState::claim_handshake(bool is_direct) {
-  if (!_processing_sem.trywait()) {
+bool HandshakeState::claim_handshake() {
+  if (!_lock.try_lock()) {
     return false;
   }
-  if (has_specific_operation(is_direct)){
-    return true;
+  // Operations are added without lock and then the poll is armed.
+  // If all handshake operations for the handshakee are finished and someone
+  // just adds an operation we may see it here. But if the handshakee is not
+  // armed yet it is not safe to procced.
+  if (has_operation_for_processor()) {
+    if (SafepointMechanism::local_poll_armed(_handshakee)) {
+      return true;
+    }
   }
-  _processing_sem.signal();
+  _lock.unlock();
   return false;
 }
 
-HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* op) {
-  bool is_direct = op->is_direct();
-
-  if (!has_specific_operation(is_direct)){
+HandshakeState::ProcessResult HandshakeState::try_process(HandshakeOperation* match_op) {
+  if (!has_operation()) {
     // JT has already cleared its handshake
-    return _no_operation;
+    return HandshakeState::_no_operation;
   }
 
   if (!possibly_can_process_handshake()) {
     // JT is observed in an unsafe state, it must notice the handshake itself
-    return _not_safe;
+    return HandshakeState::_not_safe;
   }
 
-  // Claim the semaphore if there still an operation to be executed.
-  if (!claim_handshake(is_direct)) {
-    return _state_busy;
+  // Claim the mutex if there still an operation to be executed.
+  if (!claim_handshake()) {
+    return HandshakeState::_claim_failed;
   }
 
-  // Check if the handshake operation is the same as the one we meant to execute. The
-  // handshake could have been already processed by the handshakee and a new handshake
-  // by another JavaThread might be in progress.
-  if (is_direct && op != _operation_direct) {
-    _processing_sem.signal();
-    return _no_operation;
-  }
-
-  // If we own the semaphore at this point and while owning the semaphore
+  // If we own the mutex at this point and while owning the mutex
```

Comment:
grammar - s/owning the mutex/owning the mutex we/

Reply:
Fixed